salt/_modules/metalk8s_checks.py

```diff
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+
+            # We expect nothing to be listening on this port
+            if process is None:
+                if process_on_port:
+                    # Failure:
+                    # - we expect nothing listening on this port
+                    # - a process listen on every IPs
+                    # - something already listening on the expected IP
+                    # NOTE: Normaly if a process listen on `0.0.0.0` we do not
+                    # have any other process on this port
+                    if not ip:
+                        error_process = process_on_port
+                        match = False
+                    if "0.0.0.0" in process_on_port:
+                        error_process["0.0.0.0"] = process_on_port["0.0.0.0"]
+                        match = False
+                    if ip in process_on_port:
+                        error_process[ip] = process_on_port[ip]
+                        match = False
+
+            # We expect "<process>" to be listening on this port
+            # NOTE: if nothing listening it's a failure
+            else:
+                # Failure:
+                # - nothing listening on this ip:port
+                # - nothing listening on the expected IP
+                # - something not expected already listening
+                if (
+                    not process_on_port
+                    or ip
+                    and "0.0.0.0" not in process_on_port
+                    and ip not in process_on_port
+                ):
+                    match = False
+                elif "0.0.0.0" in process_on_port and not re.match(
+                    process, process_on_port["0.0.0.0"]["name"]
+                ):
+                    error_process["0.0.0.0"] = process_on_port["0.0.0.0"]
+                    match = False
+                elif (
+                    ip
+                    and ip in process_on_port
+                    and not re.match(process, process_on_port[ip]["name"])
+                ):
+                    error_process[ip] = process_on_port[ip]
+                    match = False
+                elif not ip:
+                    for proc_ip, proc in process_on_port.items():
+                        if not re.match(process, proc["name"]):
+                            error_process[proc_ip] = proc
+                            match = False
+
+            # This "process" match what we expect
+            if match:
+                success = True
+                break
+
+        if not success:
+            fail_msg = "{} should be listening on {} but {}.".format(
+                matching.get(
+                    "description",
+                    " or ".join(process or "nothing" for process in processes),
+                ),
+                address,
+                "nothing listening"
+                if not error_process
+                else " and ".join(
+                    "{}({}) listen on {}".format(proc["name"], proc["pid"], proc_ip)
```

Comment:
I'd suggest the following, to improve (slightly) the clarity of such messages:

```python
"'{p[name]}' (PID: {p[pid]}) listens on {addr}".format(p=proc, addr=proc_ip)
```

Most important part in this suggestion is `s/listen/listens/`